src/lib/Microsoft.Health.Extensions.Fhir.R4/Service/ResourceManagementService.cs

```diff
@@ -69,13 +71,36 @@ protected async Task<TResource> CreateResourceByIdentityAsync<TResource>(Model.I
 
             propertySetter?.Invoke(resource, identifier);
 
-            return await FhirService.CreateResourceAsync(resource).ConfigureAwait(false);
+            // Generate id programatically and issue an update to FHIR service
+            // The resource will be created if the resource with the given id does not already exist.
+            resource.Id = ComputeSha256(identifier);
+            return await FhirService.UpdateResourceAsync<TResource>(resource).ConfigureAwait(false);
         }
 
         private static Model.Identifier BuildIdentifier(string value, string system)
         {
             var identifier = new Model.Identifier { Value = value, System = string.IsNullOrWhiteSpace(system) ? null : system };
             return identifier;
         }
+
+        private static string ComputeSha256(Model.Identifier identifier)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(identifier.Value, nameof(identifier.Value));
+
+            string plainTextSystemAndId = $"{identifier.System}_{identifier.Value}";
+
+            using (SHA256 hashAlgorithm = SHA256.Create())

```

Comment:
Upon reflecting on this more my main concern with using a hash algorithm is if a conflict does happen it is pretty devastating.  In such a scenario we have merged patients and combined their data resulting in a potential patient safety issue.

While not ideal, I think the better approach is to avoid a hash but instead do the lossless opaque transform and throw an error if the the result exists the 64 character length.  Long term we can work with the FHIR team to improve the conditional create/update story and transition to that when it is available.

Thoughts?

Reply:
Followed up offline.  The probability of this actually happening is to the point of actually being impossible. The probability of a conflict is around 2.69*10^-58.